gql/sort.go

```diff
@@ -0,0 +1,143 @@
+package gql
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/GannettDigital/graphql/language/ast"
+)
+
+const (
+	ascending  = "ASC"
+	descending = "DESC"
+)
+
+type lessFunc func(i, j int) bool
+
+type sortParameters struct {
+	field string
+	order string
+}
+
+// parseSortParameters parses the given argument returning a the sort parameters.
+// If the argument is nil the returned value is nil.
+func parseSortParameters(arg interface{}) (*sortParameters, error) {
+	if arg == nil {
+		return nil, nil
+	}
+	fields, ok := arg.([]*ast.ObjectField)
+	if !ok {
+		return nil, errors.New("unable to parse sort argument")
+	}
+
+	var params sortParameters
+	for _, f := range fields {
+		switch f.Name.Value {
+		case "Field":
+			v, ok := f.GetValue().(*ast.StringValue)
+			if !ok {
+				return nil, errors.New("unable to parse sort argument field Field")
+			}
+			params.field = v.Value
+		case "Order":
+			v, ok := f.GetValue().(*ast.StringValue)
+			if !ok {
+				return nil, errors.New("unable to parse sort argument field Order")
+			}
+			switch v.Value {
+			case ascending, descending, "":
+			default:
+				return nil, fmt.Errorf("sort order must be %q or %q or undefined", ascending, descending)
+			}
+			params.order = v.Value
+		}
+	}
+
+	return &params, nil
+}
+
+// listSort will sort the given list in place according to the params specified.
+// The type of the specified field is essential information when sorting but can't be determined until the list to be
+// sorted is available which is why this does not follow the golang standard sort interface.
+// An error can occur if sort the fields in the list items is not consistently the same type or an unsupported type.
+func listSort(params *sortParameters, list []interface{}) error {
+	if len(list) < 2 {
+		return nil
+	}
+
+	errChan := make(chan error)
+
+	go unprotectedListSort(params, list, errChan)
```

Comment:
Is there a reason we are goroutining if we are blocking right below it?

Reply:
I guess to recover? But I don't see how the less func could encounter a bad type if we are erroring out on non float string or ints

Reply:
Yes it is because of the recover. The only way it will panic is if the list a `[]interface{}` has list items with varied types. Though that should be a really rare case I still wanted to handle it. This is a good example of why to avoid empty interfaces where possible.